### ISSUE TITLE
Implement 2fa for logging in

### ISSUE THREAD

**Issue by MiniDigger** [thelounge/lounge#842]

**Comment by crdnl** [comment]
Both TOTP and U2F would be a good starting point and it covers 99% of use cases.

**Comment by McInkay** [comment]
As nameless on IRC was saying they will pick this up, I thought I'd detail what needs to be done for this. Presuming just TOTP at this point (simplest and by far the most commonly used 2fa in my experience).

1. Using something like https://www.npmjs.com/package/otplib
2. Button to generate secret and show QR code
3. Ensure code is scanned properly (get user to input code, and check it is valid)
4. Store secret somewhere on the server (biggest issue, we don't currently have a way to store secrets on the server ... should we encrypt using user's password, or leave in plain text?)
5. Inside [auth](https://github.com/thelounge/thelounge/blob/master/src/plugins/auth/local.js), send client that we need OTP (if password was correct, and user has OTP set up).
6. Client shows input for OTP, sends back to server
7. Check code is correct, log in if successful, send error back if not (and cycle back to 6)

Have I missed anything?

**Comment by McInkay** [comment]
> Said button should be aware if you have 2FA setup already, and if you do, it would regenerate the existing key.

Yeah, either that or we don't show the button after you've set up 2FA. I know it's quite common to only allow you to see the QR code once. I'm not sure how I feel about that though. Thoughts?

> What about the CLI command that changes user password, should that remove 2FA (by asking perhaps)?

Good question. I'm leaning towards either "don't have anything to do with 2FA in that command" or "it's an option to change 2FA". But I'm not sure we should remove it entirely.

**Comment by xPaw** [comment]
>  I know it's quite common to only allow you to see the QR code once.

Show? Yeah, you only do that once. But we should allow re-generating it (e.g if you want to move 2fa apps or whatever).

> But I'm not sure we should remove it entirely.

Might be complicated if admin changes 2FA to communicate that to the user.

**Comment by McInkay** [comment]
> Show? Yeah, you only do that once. But we should allow re-generating it (e.g if you want to move 2fa apps or whatever).

Oh sorry, I misunderstood what you were saying. I thought you were saying we should show the existing key, Apologies.

We probably want something more complicated than 1 single button press to regenerate the secret and overwrite it though. I guess we could not overwrite it until the user has input a code from the new secret?

> Might be complicated if admin changes 2FA to communicate that to the user.

Yeah, good point. That feels really dodgy as well. I think we should leave it out entirely. The admin can manually edit the user's config to remove the 2fa secret if they absolutely need to. I don't think we want to officially support something otherwise.

**Comment by lord2y** [comment]
> As nameless on IRC was saying they will pick this up, I thought I'd detail what needs to be done for this. Presuming just TOTP at this point (simplest and by far the most commonly used 2fa in my experience).
> 
>     1. Using something like https://www.npmjs.com/package/otplib
> 
>     2. Button to generate secret and show QR code
> 
>     3. Ensure code is scanned properly (get user to input code, and check it is valid)
> 
>     4. Store secret somewhere on the server (biggest issue, we don't currently have a way to store secrets on the server ... should we encrypt using user's password, or leave in plain text?)
> 
>     5. Inside [auth](https://github.com/thelounge/thelounge/blob/master/src/plugins/auth/local.js), send client that we need OTP (if password was correct, and user has OTP set up).
> 
>     6. Client shows input for OTP, sends back to server
> 
>     7. Check code is correct, log in if successful, send error back if not (and cycle back to 6)
> 
> 
> Have I missed anything?

Print recovery code(s) during the 2fa activation phase in case it won't be possible to disable the 2fa from the cli.

**Comment by strugee** [comment]
> Store secret somewhere on the server (biggest issue, we don't currently have a way to store secrets on the server ... should we encrypt using user's password, or leave in plain text?)

I don't think this is actually a problem? What's the threat model? If it's "I don't trust my server admin so I want to hide this info from them" you're out of luck anyway for other reasons, and if it's "someone might compromise the server and steal my 2FA credentials", the 2FA creds don't get the attacker access to anything they wouldn't already have by compromising the server (this isn't true for passwords because users reuse passwords all the time, but they can't do that for TOTP). And if the server is compromised you'd probably want to rotate all that stuff anyway just in case.

I might be missing something though.

**Comment by igalic** [comment]
> What's the threat model?

that's always the most important question when implementing any security related features

once the threat model is decided, it should be documented along with the implementation.

**Comment by crdnl** [comment]
> Print recovery code(s) during the 2fa activation phase in case it won't be possible to disable the 2fa from the cli.

Generally, the recovery codes are still accessible after 2FA setup. Ex. Google gives access to recovery codes, and allows regen-ing them.

> > Might be complicated if admin changes 2FA to communicate that to the user.
> 
> Yeah, good point. That feels really dodgy as well. I think we should leave it out entirely. The admin can manually edit the user's config to remove the 2fa secret if they absolutely need to. I don't think we want to officially support something otherwise.

Maybe, instead of using the standard password reset in the CLI, to move it to a seperate sub-command.

**Comment by skapytek** [comment]
Any updates on adding 2FA?

**Comment by Staubgeborener** [comment]
> Any updates on adding 2FA?

This issue is 10 years old and we are already in 2026 where 2FA is not a default feature for thelounge. Instead, asking for this state of the art feature leads into a spam flag.


Wow.
